components/support/rc_crypto/nss/src/ec.rs

```diff
@@ -0,0 +1,378 @@
+    ptr,
+};
+
+#[derive(Serialize, Deserialize, Clone, Copy, Debug, PartialEq)]
+#[repr(u8)]
+pub enum Curve {
+    P256,
+}
+const CRV_P256: &str = "P-256";
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub struct EcKey {
+    curve: String,
+    // The `d` value of the EC Key.
+    private_key: Vec<u8>,
+    // The uncompressed x,y-representation of the public component of the EC Key.
+    public_key: Vec<u8>,
+}
+
+impl EcKey {
+    pub fn new(curve: Curve, private_key: &[u8], public_key: &[u8]) -> Self {
+        let curve = match curve {
+            Curve::P256 => CRV_P256,
+        };
+        Self {
+            curve: curve.to_owned(),
+            private_key: private_key.to_vec(),
+            public_key: public_key.to_vec(),
+        }
+    }
+
+    pub fn from_coordinates(curve: Curve, d: &[u8], x: &[u8], y: &[u8]) -> Result<Self> {
+        let ec_point = create_ec_point_for_coordinates(x, y)?;
+        Ok(EcKey::new(curve, d, &ec_point))
+    }
+
+    pub fn curve(&self) -> Curve {
+        if self.curve == CRV_P256 {
+            return Curve::P256;
+        }
+        unimplemented!("It is impossible to create a curve object with a different CRV.")
+    }
+
+    pub fn public_key(&self) -> &[u8] {
+        &self.public_key
+    }
+
+    pub fn private_key(&self) -> &[u8] {
+        &self.private_key
+    }
+}
+
+fn create_ec_point_for_coordinates(x: &[u8], y: &[u8]) -> Result<Vec<u8>> {
+    if x.len() != y.len() {
+        return Err(ErrorKind::InternalError.into());
+    }
+    let mut buf = vec![0u8; x.len() + y.len() + 1];
+    buf[0] = u8::try_from(nss_sys::EC_POINT_FORM_UNCOMPRESSED)?;
+    let mut offset = 1;
+    buf[offset..offset + x.len()].copy_from_slice(x);
+    offset += x.len();
+    buf[offset..offset + y.len()].copy_from_slice(y);
+    Ok(buf)
+}
+
+pub fn generate_keypair(curve: Curve) -> Result<(PrivateKey, PublicKey)> {
+    ensure_nss_initialized();
+    // 1. Create EC params
+    let params_buf = create_ec_params_for_curve(curve)?;
+    let mut params = nss_sys::SECItem {
+        type_: nss_sys::SECItemType::siBuffer,
+        data: params_buf.as_ptr() as *mut c_uchar,
+        len: c_uint::try_from(params_buf.len())?,
+    };
+
+    // 2. Generate the key pair
+    // The following code is adapted from:
+    // https://searchfox.org/mozilla-central/rev/f46e2bf881d522a440b30cbf5cf8d76fc212eaf4/dom/crypto/WebCryptoTask.cpp#2389
+    let mech = match curve {
+        Curve::P256 => nss_sys::CKM_EC_KEY_PAIR_GEN,
+    };
+    let slot = slot::get_internal_slot()?;
+    let mut pub_key: *mut nss_sys::SECKEYPublicKey = ptr::null_mut();
+    let prv_key = PrivateKey::from(curve, unsafe {
+        PK11PrivateKey::from_ptr(nss_sys::PK11_GenerateKeyPair(
+            slot.as_mut_ptr(),
+            mech.into(),
+            &mut params as *mut _ as *mut c_void,
+            &mut pub_key,
+            nss_sys::PR_FALSE,
+            nss_sys::PR_FALSE,
+            ptr::null_mut(),
```

Comment:
Perhaps for a followup...is there any way to pass these as named fields rather than having to remember what all 7 arguments to `PK11_GenerateKeyPair` are?